cpp/src/copying/contiguous_split.cu

```diff
@@ -14,465 +14,923 @@
-                 [mr, stream, &device_split_info](table_view const& t) {
-                   return alloc_and_copy(t, device_split_info, mr, stream);
+  // if inputs are empty, just return num_partitions empty tables
+  if (input.column(0).size() == 0) {
+    std::vector<contiguous_split_result> result;
+    result.reserve(num_partitions);
+
+    auto iter = thrust::make_counting_iterator(0);
+    std::transform(
+      iter, iter + num_partitions, std::back_inserter(result), [&input](int partition_index) {
+        return contiguous_split_result{input, std::make_unique<rmm::device_buffer>()};
+      });
+
+    return std::move(result);
+  }
+
+  size_t const num_root_columns = input.num_columns();
+
+  // compute # of source buffers (column data, validity, children), # of partitions
+  // and total # of buffers
+  size_type const num_src_bufs = count_src_bufs(input.begin(), input.end());
+  size_t const num_bufs        = num_src_bufs * num_partitions;
+
+  // clang-format off
+  // packed block of memory 1. split indices and src_buf_info structs
+  size_t const indices_size = cudf::util::round_up_safe((num_partitions + 1) * sizeof(size_type), split_align);
+  size_t const src_buf_info_size = cudf::util::round_up_safe(num_src_bufs * sizeof(src_buf_info), split_align);
+  // host-side
+  std::vector<uint8_t> h_indices_and_source_info(indices_size + src_buf_info_size);
+  size_type* h_indices = reinterpret_cast<size_type*>(h_indices_and_source_info.data());
+  src_buf_info* h_src_buf_info = reinterpret_cast<src_buf_info*>(h_indices_and_source_info.data() + indices_size);
+  // device-side
+  // gpu-only : stack space needed for nested list offset calculation
+  int const offset_stack_partition_size = compute_offset_stack_size(input.begin(), input.end());
+  size_t const offset_stack_size = offset_stack_partition_size * num_partitions * sizeof(size_type);
+  rmm::device_buffer d_indices_and_source_info(indices_size + src_buf_info_size + offset_stack_size, stream, rmm::mr::get_current_device_resource());
+  size_type* d_indices         = reinterpret_cast<size_type*>(d_indices_and_source_info.data());
+  src_buf_info* d_src_buf_info = reinterpret_cast<src_buf_info*>(reinterpret_cast<uint8_t*>(d_indices_and_source_info.data()) + indices_size);
+  size_type* d_offset_stack =    reinterpret_cast<size_type*>(reinterpret_cast<uint8_t*>(d_indices_and_source_info.data()) +
+                                 indices_size + src_buf_info_size);
+  // clang-format on
+
+  // compute splits -> indices.
+  h_indices[0]              = 0;
+  h_indices[num_partitions] = input.column(0).size();
+  std::copy(splits.begin(), splits.end(), std::next(h_indices));
+
+  // setup source buf info
+  setup_source_buf_info(input.begin(), input.end(), h_src_buf_info, h_src_buf_info);
+
+  // HtoD indices and source buf info to device
+  CUDA_TRY(cudaMemcpyAsync(
+    d_indices, h_indices, indices_size + src_buf_info_size, cudaMemcpyHostToDevice, stream));
```

Comment:
Need to get the value when it's a `rmm::cuda_stream_view`. (No implicit cast to cudaStream_t).

```suggestion
    d_indices, h_indices, indices_size + src_buf_info_size, cudaMemcpyHostToDevice, stream.value()));
```